snowfall/decoding/util.py

```diff
@@ -0,0 +1,45 @@
+# Copyright (c)  2021  Xiaomi Corp.       (author: Fangjun Kuang)
+
+import k2
+import torch
+
+
+def get_log_probs(phone_fsas: k2.Fsa, nnet_output: torch.Tensor,
+                  len_per_path: torch.Tensor):
+    '''
+    Args:
+      phone_fsas:
+        An FsaVec with three axes [path][state][arc]
+      nnet_output:
+        A 3-D torch.Tensor (num_path, seq_len, num_phones+1)
+      len_per_path:
+        A 1-D torch.Tensor of shape (num_path)
+    '''
+    assert len(phone_fsas.shape) == 3
+    offset = phone_fsas.arcs.row_splits(2)[phone_fsas.arcs.row_splits(
+        1).long()]
+    expected_len_per_path = offset[1:] - offset[:-1]
+    assert torch.all(torch.eq(len_per_path, expected_len_per_path.cpu()))
+
+    phone_fsas = phone_fsas.to('cpu')
+    nnet_output = nnet_output.to('cpu')
+    assert phone_fsas.arcs.dim0() == nnet_output.shape[0]
+
+    probs = []
+    num_paths = phone_fsas.arcs.dim0()
+    for i in range(num_paths):
+        this_fsa = k2.index(phone_fsas, torch.tensor([i], dtype=torch.int32))
+        len_this_fsa = len_per_path[i]
+        this_fsa_nnet_output = nnet_output[i]
+        this_prob = []
+        labels = this_fsa.labels
+        for idx, row in enumerate(this_fsa_nnet_output):
+            if idx >= len_this_fsa:
+                break
+            this_prob.append(row[labels[idx]].item())
```

Comment:
Should there be a +1 here?  I thought there was a shift, to map -1 to 0.

Reply:
.. also, I'm not sure if we may have a problem at the start of the sequence, due to there being no epsilon there??
I'm not sure where `phone_fsas` comes from, i.e. whether there were epsilons between each phone before adding the epsilon self-loops (I assume not), and whether there are epsilons at the start and/or the end in the sequence given as input to the network.

Reply:
> Should there be a +1 here? I thought there was a shift, to map -1 to 0.

The input to the second pass model shifts `-1` (i.e., EOS) to `0`. The output of the second pass model contains
only blank + phone_ids, no EOS.

I should have skipped the last label `-1` of every path since there is no corresponding output for it.

Reply:
> I'm not sure where phone_fsas comes from, i.e. whether there were epsilons between each phone before adding the epsilon self-loops (I assume not)

The `phone_fsas` is created following https://github.com/k2-fsa/k2/issues/641#issuecomment-770548577
For example, if the `phone_seqs` is [c, a, t], then `phone_fsas` is [0, c, 0, a, 0, t, 0, -1].

Before adding the epsilon self-loop, there were no epsilons between phones. It follows
the comment in https://github.com/k2-fsa/k2/issues/641#issue-796593136
```python
     phone_seqs = k2.index(mbr_lats.phones, paths)
     # Remove epsilons from `phone_seqs`
     phone_seqs = k2.ragged.remove_values_eq(phones, 0)
```

> , and whether there are epsilons at the start and/or the end in the sequence given as input to the network.

For the second pass network, there is an epsilon at the start of each sequence and there is an EOS at
the end of each sequence.

Reply:
mm, there must be some shift for which those probabilities are all close to zero.  Perhaps if you print out the best-path label sequences, including epsilons, from the second pass it would be clear what it is doing, e.g. does it look like [ 0 nonzero 0 nonzero ... ?]

Reply:
> does it look like [ 0 nonzero 0 nonzero ... ?]

Thanks, will check it.